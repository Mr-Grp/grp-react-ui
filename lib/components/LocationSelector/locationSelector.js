var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import React, { useState, useEffect, useCallback } from 'react';
import { defaultLocation } from './config';
var LocationSelector = function (_a) {
    var onSelectProvince = _a.onSelectProvince, onSelectCity = _a.onSelectCity, defaultProvince = _a.defaultProvince, defaultCity = _a.defaultCity, _b = _a.className, className = _b === void 0 ? '' : _b, _c = _a.style, style = _c === void 0 ? {} : _c, _d = _a.locationData, locationData = _d === void 0 ? defaultLocation : _d, locationSearchMethod = _a.locationSearchMethod, citySearchMethod = _a.citySearchMethod, cityValue = _a.cityValue, provinceValue = _a.provinceValue, _e = _a.title, title = _e === void 0 ? '地区选择器' : _e;
    var _f = useState([]), locationOptions = _f[0], setLocationOptions = _f[1];
    var _g = useState([]), cityOptions = _g[0], setCityOptions = _g[1];
    var _h = useState(defaultCity), _cityId = _h[0], setCityId = _h[1];
    var _j = useState(defaultProvince), _provinceId = _j[0], setProvinceId = _j[1];
    var _k = useState(false), loading = _k[0], setLoading = _k[1];
    // 受控组件
    var cityId = cityValue !== undefined ? cityValue : _cityId;
    var provinceId = provinceValue !== undefined ? provinceValue : _provinceId;
    // 获取地域和省份
    var getProvinces = useCallback(function () { return __awaiter(void 0, void 0, void 0, function () {
        var result;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    setLoading(true);
                    return [4 /*yield*/, locationSearchMethod()];
                case 1:
                    result = _a.sent();
                    setLocationOptions(result);
                    setLoading(false);
                    return [2 /*return*/];
            }
        });
    }); }, [locationSearchMethod]);
    // 获取城市选项列表
    var getCities = useCallback(function (id) { return __awaiter(void 0, void 0, void 0, function () {
        var result;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!!citySearchMethod) return [3 /*break*/, 1];
                    // 没有搜索接口，使用选中省份的 children
                    locationOptions === null || locationOptions === void 0 ? void 0 : locationOptions.forEach(function (item) {
                        var _a;
                        (_a = item.children) === null || _a === void 0 ? void 0 : _a.forEach(function (p) {
                            if (p.id === id) {
                                setCityOptions(p.children || []);
                            }
                        });
                    });
                    return [3 /*break*/, 3];
                case 1:
                    // 使用接口
                    setLoading(true);
                    return [4 /*yield*/, citySearchMethod(id)];
                case 2:
                    result = _a.sent();
                    setCityOptions(result);
                    setLoading(false);
                    _a.label = 3;
                case 3: return [2 /*return*/];
            }
        });
    }); }, [citySearchMethod, locationOptions]);
    // 初始化生命周期
    useEffect(function () {
        if (!locationSearchMethod) {
            // 使用默认数据
            setLocationOptions(locationData);
        }
        else {
            // 使用接口数据
            getProvinces();
        }
    }, [locationSearchMethod, locationData, getProvinces]);
    // province 发生改变的时候，获取新 city 列表，通知父组件
    useEffect(function () {
        if (provinceId) {
            getCities(provinceId);
        }
    }, [provinceId, getCities]);
    // 选择省份
    var _onSelectProvince = function (nextProvinceId) { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            if (nextProvinceId !== provinceId) {
                // 新的 province，同时清空选中的 city
                setProvinceId(nextProvinceId);
                setCityId(undefined);
                onSelectProvince && onSelectProvince(nextProvinceId);
                onSelectCity && onSelectCity(undefined);
            }
            else {
                // 取消选择，同时清空城市
                setProvinceId(undefined);
                setCityId(undefined);
                setCityOptions([]);
                onSelectProvince && onSelectProvince(undefined);
                onSelectCity && onSelectCity(undefined);
            }
            return [2 /*return*/];
        });
    }); };
    // 选择城市
    var _onSelectCity = function (nextCityId) {
        if (nextCityId !== cityId) {
            // 新的city
            setCityId(nextCityId);
            onSelectCity && onSelectCity(nextCityId);
        }
        else {
            // 取消选择
            setCityId(undefined);
            onSelectCity && onSelectCity(undefined);
        }
    };
    return (React.createElement("div", { className: "g-location-selector ".concat(className), style: style },
        loading && React.createElement("div", { className: "g-loading" }, "\u52A0\u8F7D\u4E2D..."),
        React.createElement("div", { className: "g-location-selector-title" }, title),
        React.createElement("div", { className: "g-location" }, locationOptions.map(function (location) {
            var _a, _b;
            return (React.createElement("div", { className: "g-location-item", key: location.id },
                React.createElement("div", { className: "g-location-item-label" },
                    location.name,
                    ":"),
                React.createElement("div", { className: "g-province" },
                    React.createElement("div", { className: "g-province-list" }, (_a = location.children) === null || _a === void 0 ? void 0 : _a.map(function (province) {
                        return (React.createElement("div", { key: province.id, className: "".concat(province.id === provinceId ? 'g-province-item-selected' : 'g-province-item'), onClick: function () { return _onSelectProvince(province.id); } }, province.name));
                    })),
                    React.createElement("div", { className: "g-city" }, ((_b = location.children) === null || _b === void 0 ? void 0 : _b.find(function (province) { return province.id === provinceId && !!cityOptions.length; }))
                        && (React.createElement("div", { className: "g-city-list" }, cityOptions.map(function (city) {
                            return (React.createElement("div", { key: city.id, className: " ".concat(city.id === cityId ? 'g-city-item-selected' : 'g-city-item'), onClick: function () { return _onSelectCity(city.id); } }, city.name));
                        })))))));
        }))));
};
export default LocationSelector;
