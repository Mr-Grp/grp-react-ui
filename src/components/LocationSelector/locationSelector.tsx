import React, { useState, useEffect, useCallback, CSSProperties } from 'react';
import { defaultLocation } from './config';

// 地区的格式
export interface ILocationItem {
  id: number;
  name: string;
  children?: ILocationItem[] | null;
}

/**
 * 地区选择器
 * onSelectProvince 省份被选择时触发
 * onSelectCity 城市被选择时触发
 * defaultCity 默认选中 city
 * defaultProvince 默认选中 province
 * cityValue 选中 city 受控组件
 * provinceValue 选中 province 受控组件
 * title 标题
 * className 最外层类名
 * style 最外层样式
 * locationData 组件数据，不传入采用默认数据
 * locationSearchMethod 获取地区省份的方法，不传入采用默认数据
 * citySearchMethod 获取城市的方法，不传入在所在省份数据的 children 查找
 *
 * 注意：省份列表必须分为两层：地域和省份  [{id: 1, name: '华北’, children: [{id：2， name: '河北省', children: []}]}],在使用 locationData 和 locationSearchMethod 注意！！！！！！
 */

export interface ILocationSelectorProps {
  onSelectProvince?: (id: number | undefined) => void;
  onSelectCity?: (id: number | undefined) => void;
  defaultProvince?: number;
  defaultCity?: number;
  provinceValue?: number;
  cityValue?: number;
  title?: string;
  className?: string;
  style?: CSSProperties;
  locationData?: ILocationItem[];
  locationSearchMethod?: () => ILocationItem[] | Promise<ILocationItem[]>;
  citySearchMethod?: (id: number) => ILocationItem[] | Promise<ILocationItem[]>;
}

const LocationSelector: React.FC<ILocationSelectorProps> = ({
  onSelectProvince,
  onSelectCity,
  defaultProvince,
  defaultCity,
  className = '',
  style = {},
  locationData = defaultLocation,
  locationSearchMethod,
  citySearchMethod,
  cityValue,
  provinceValue,
  title = '地区选择器',
}) => {
  const [locationOptions, setLocationOptions] = useState<ILocationItem[]>([]);
  const [cityOptions, setCityOptions] = useState<ILocationItem[]>([]);
  const [_cityId, setCityId] = useState<number | undefined>(defaultCity);
  const [_provinceId, setProvinceId] = useState<number | undefined>(defaultProvince);
  const [loading, setLoading] = useState<boolean>(false);

  // 受控组件
  const cityId = cityValue !== undefined ? cityValue : _cityId;
  const provinceId = provinceValue !== undefined ? provinceValue : _provinceId;
  // 获取地域和省份
  const getProvinces = useCallback(async () => {
    setLoading(true);
    const result: any = await locationSearchMethod!();
    setLocationOptions(result);
    setLoading(false);
  }, [locationSearchMethod]);

  // 获取城市选项列表
  const getCities = useCallback(async (id: number) => {
    if (!citySearchMethod) {
      // 没有搜索接口，使用选中省份的 children
      locationOptions?.forEach(item => {
        item.children?.forEach(p => {
          if (p.id === id) {
            setCityOptions(p.children || []);
          }
        });
      });
    } else {
      // 使用接口
      setLoading(true);
      const result: any = await citySearchMethod!(id);
      setCityOptions(result);
      setLoading(false);
    }
  }, [citySearchMethod, locationOptions]);

  // 初始化生命周期
  useEffect(() => {
    if (!locationSearchMethod) {
      // 使用默认数据
      setLocationOptions(locationData);
    } else {
      // 使用接口数据
      getProvinces();
    }
  }, [locationSearchMethod, locationData, getProvinces]);

  // province 发生改变的时候，获取新 city 列表，通知父组件
  useEffect(() => {
    if (provinceId) {
      getCities(provinceId);
    }
  }, [provinceId, getCities]);

  // 选择省份
  const _onSelectProvince = async (nextProvinceId: number) => {
    if (nextProvinceId !== provinceId) {
      // 新的 province，同时清空选中的 city
      setProvinceId(nextProvinceId);
      setCityId(undefined);
      onSelectProvince && onSelectProvince(nextProvinceId);
      onSelectCity && onSelectCity(undefined);
    } else {
      // 取消选择，同时清空城市
      setProvinceId(undefined);
      setCityId(undefined);
      setCityOptions([]);
      onSelectProvince && onSelectProvince(undefined);
      onSelectCity && onSelectCity(undefined);
    }
  };

  // 选择城市
  const _onSelectCity = (nextCityId: number) => {
    if (nextCityId !== cityId) {
      // 新的city
      setCityId(nextCityId);
      onSelectCity && onSelectCity(nextCityId);
    } else {
      // 取消选择
      setCityId(undefined);
      onSelectCity && onSelectCity(undefined);
    }
  };

  return (
    <div className={`g-location-selector ${className}`} style={style}>
      {loading && <div className="g-loading">加载中...</div>}
      <div className="g-location-selector-title">{title}</div>
      <div className="g-location">
        {locationOptions.map(location => {
          return (
            <div className="g-location-item" key={location.id}>
              <div className="g-location-item-label">
                {location.name}:
              </div>
              <div className="g-province">
                <div className="g-province-list">
                  {location.children?.map(province => {
                    return (
                      <div
                        key={province.id}
                        className={`${province.id === provinceId ? 'g-province-item-selected' : 'g-province-item'}`}
                        onClick={() => _onSelectProvince(province.id)}
                      >
                        {province.name}
                      </div>
                    );
                  })}
                </div>
                <div className="g-city">
                  {location.children?.find(province => province.id === provinceId && !!cityOptions.length)
                    && (
                      <div className="g-city-list">
                        {cityOptions.map(city => {
                          return (
                            <div
                              key={city.id}
                              className={` ${city.id === cityId ? 'g-city-item-selected' : 'g-city-item'}`}
                              onClick={() => _onSelectCity(city.id)}
                            >
                              {city.name}
                            </div>
                          );
                        })}
                      </div>
                    )}
                </div>
              </div>
            </div>
          );
        })}
      </div>
    </div>
  );
};

export default LocationSelector;
