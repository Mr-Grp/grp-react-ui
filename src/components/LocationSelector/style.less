@g-location-province-height: 36px;
@g-location-city-height: 14px;
@g-location-theme: @orange;

.g-location-selector {
  min-width: 750px;
  padding: 24px;
  border: 1px solid @gray-200;
  border-radius: 5px;
  position: relative;
  .g-loading {
    position: absolute;
    left: 0;
    right: 0;
    bottom: 0;
    top: 0;
    display: flex;
    align-items: center;
    justify-content: center;
    background-color: rgba(255, 255, 255, 0.6);
    color: @g-location-theme;
    font-size: @font-size-sm;
  }

  .g-location-selector-title {
    font-size: @font-size-base;    
    font-weight: @font-weight-bold;
    color: @gray-800;
    margin-bottom: 16px;
  }

  .g-location {
    .g-location-item {
      display: flex;
      margin-bottom: 8px;
      .g-location-item-label {
        flex: 0 0 33px;
        font-size: @font-size-sm;
        font-family: @font-family-sans-serif;
        font-weight: @font-weight-normal;
        color: @gray-500;
        line-height: @g-location-province-height;
      }
      .g-province {
        flex: 1;
        margin-left: 22px;
        .g-province-list {
          display: flex;
          flex-wrap: wrap;
          .g-province-item,
          .g-province-item-selected {
            cursor: pointer;
            margin-right: 5px;
            padding: 0 16px;
            height: @g-location-province-height;
            line-height: @g-location-province-height;
            transition: all 0.3s;
            font-size: @font-size-sm;
            font-family: @font-family-sans-serif;
            font-weight: @font-weight-normal;
            color: @gray-800;
            border-radius: 2px;
          }
          .g-province-item-selected,
          .g-province-item:hover {
            color: @white;
            background: @g-location-theme;
          }
        }
        .g-city {
          .g-city-list {
            display: flex;
            flex-wrap: wrap;
            background: #f4f4f4;
            border-radius: 2px;
            padding: 8px 7px;
            margin-top: 9px;
            min-height: 46px;
    
            .g-city-item,
            .g-city-item-selected {
              cursor: pointer;
              margin: 8px 15px;
              height: @g-location-city-height;
              line-height: @g-location-city-height;
              transition: all 0.3s;
              font-size: @font-size-sm;
              font-family: @font-family-sans-serif;
              font-weight: @font-weight-normal;
              color: @gray-600;
            }
            .g-city-item-selected,
            .g-city-item:hover {
              color: @g-location-theme;
            }
          }
        }
      }
    }
  }
}
