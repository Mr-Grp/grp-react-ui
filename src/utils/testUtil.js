import { act } from 'react-dom/test-utils';

export const findTestWrapper = (wrapper, tag) => {
  return wrapper.find(`[data-test="${tag}"]`);
};

export const waitForComponentToPaint = async (wrapper) => {
  await act(async () => {
    await new Promise(resolve => {
      setTimeout(resolve, 0);
    });
    wrapper.update();
  });
};
