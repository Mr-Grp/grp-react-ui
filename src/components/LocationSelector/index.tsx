import LocationSelector from './locationSelector';
import './style.less';

export default LocationSelector;
