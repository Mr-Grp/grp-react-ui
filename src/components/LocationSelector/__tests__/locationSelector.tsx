import { shallow, mount } from 'enzyme';
import { waitForComponentToPaint } from '../../../utils/testUtil';
import LocationSelector from '..';
import { defaultLocation } from '../config';

describe('LocationSelector 测试', () => {
  test('可以正确出现的在页面里', () => {
    const wrapper = shallow(
      <LocationSelector />,
    );
    expect(wrapper.find('.g-location-selector')).toExist();
  });

  test('默认使用config数据', () => {
    const wrapper = mount(
      <LocationSelector />,
    );
    const locationElem = wrapper.find('.g-location-item');
    expect(locationElem.length).toEqual(defaultLocation.length);

    for (let i = 0; i < locationElem.length; i++) {
      const locationItem = locationElem.at(i);
      const locationLabelElem = locationItem.find('.g-location-item-label').text();
      const provinceElem = locationItem.find('.g-province-item');
      expect(locationLabelElem).toMatch(defaultLocation[i].name);

      const provinceData = defaultLocation[i].children;
      expect(provinceElem.length).toBe(provinceData.length);
      for (let j = 0; j < provinceElem.length; j++) {
        expect(provinceElem.at(j).text()).toMatch(provinceData[j].name);
      }
    }
  });

  test('点击 province 时会选中,并展示对应 city 列表,city 列表数据默认为 province.children', () => {
    const wrapper = mount(
      <LocationSelector />,
    );
    let provinceElem = wrapper.find('.g-province-item');
    const provinceData = defaultLocation.map(item => item.children).flat();
    expect(provinceElem.length).toBe(provinceData.length);
    const len = provinceElem.length;
    const random = Math.floor(Math.random() * len);
    const targetElem = provinceElem.at(random);
    targetElem.simulate('click');
    wrapper.update();
    provinceElem = wrapper.find('.g-province-item-selected');
    expect(provinceElem.text()).toMatch(provinceData[random].name);

    const cityData = provinceData[random].children;
    const cityElem = wrapper.find('.g-city-item');
    expect(cityElem.length).toBe(cityData.length);

    for (let i = 0; i < cityElem.length; i++) {
      expect(cityElem.at(i).text()).toMatch(cityData[i].name);
    }
  });

  test('再次点击相同 province 时,取消选中,并隐藏 city 列表', () => {
    const wrapper = mount(
      <LocationSelector />,
    );
    let provinceElem = wrapper.find('.g-province-item');
    const len = provinceElem.length;
    const random = Math.floor(Math.random() * len);
    const targetElem = provinceElem.at(random);
    targetElem.simulate('click');
    wrapper.update();
    provinceElem = wrapper.find('.g-province-item-selected');
    expect(provinceElem.length).toBe(1);
    provinceElem.simulate('click');
    wrapper.update();
    provinceElem = wrapper.find('.g-province-item-selected');
    expect(provinceElem.length).toBe(0);
  });

  test('点击 city 时会选中', () => {
    const wrapper = mount(
      <LocationSelector />,
    );
    const provinceElem = wrapper.find('.g-province-item');
    const provinceData = defaultLocation.map(item => item.children).flat();
    expect(provinceElem.length).toBe(provinceData.length);
    const len = provinceElem.length;
    let random = Math.floor(Math.random() * len);
    let targetElem = provinceElem.at(random);
    targetElem.simulate('click');
    wrapper.update();

    const cityData = provinceData[random].children;
    const cityElem = wrapper.find('.g-city-item');

    random = Math.floor(Math.random() * cityElem.length);
    targetElem = cityElem.at(random);
    targetElem.simulate('click');
    wrapper.update();
    targetElem = wrapper.find('.g-city-item-selected');
    expect(targetElem.text()).toMatch(cityData[random].name);
  });

  test('再次点击相同 city 时,取消选中', () => {
    const wrapper = mount(
      <LocationSelector />,
    );
    const provinceElem = wrapper.find('.g-province-item');
    const provinceData = defaultLocation.map(item => item.children).flat();
    expect(provinceElem.length).toBe(provinceData.length);
    const len = provinceElem.length;
    let random = Math.floor(Math.random() * len);
    let targetElem = provinceElem.at(random);
    targetElem.simulate('click');
    wrapper.update();

    const cityElem = wrapper.find('.g-city-item');

    random = Math.floor(Math.random() * cityElem.length);
    targetElem = cityElem.at(random);
    targetElem.simulate('click');
    wrapper.update();
    targetElem = wrapper.find('.g-city-item-selected');
    expect(targetElem.length).toBe(1);
    targetElem.simulate('click');
    wrapper.update();
    targetElem = wrapper.find('.g-city-item-selected');
    expect(targetElem.length).toBe(0);
  });

  test('传入 onSelectProvince, 点击 province,会调用该方法,并传入 id', () => {
    const fn = jest.fn();
    const wrapper = mount(
      <LocationSelector onSelectProvince={fn} />,
    );
    const provinceElem = wrapper.find('.g-province-item');
    const provinceData = defaultLocation.map(item => item.children).flat();
    const random = Math.floor(Math.random() * provinceElem.length);
    const targetElem = provinceElem.at(random);
    targetElem.simulate('click');
    expect(fn).toBeCalledWith(provinceData[random].id);
  });

  test('传入 onSelectCity, 点击 city, 会调用该方法,并传入 id', () => {
    const fn = jest.fn();
    const wrapper = mount(
      <LocationSelector onSelectCity={fn} />,
    );
    const provinceElem = wrapper.find('.g-province-item');
    const provinceData = defaultLocation.map(item => item.children).flat();
    let random = Math.floor(Math.random() * provinceElem.length);
    const targetElem = provinceElem.at(random);
    targetElem.simulate('click');
    wrapper.update();
    const cityData = provinceData[random].children;
    const cityElem = wrapper.find('.g-city-item');
    random = Math.floor(Math.random() * cityElem.length);
    cityElem.at(random).simulate('click');
    expect(fn).toBeCalledWith(cityData[random].id);
  });

  test('传入正确的 defaultProvince, 会设置 province 默认值', () => {
    const provinceData = defaultLocation.map(item => item.children).flat();
    const random = Math.floor(Math.random() * provinceData.length);
    const wrapper = mount(
      <LocationSelector defaultProvince={provinceData[random].id} />,
    );
    const provinceElem = wrapper.find('.g-province-item-selected');
    expect(provinceElem.text()).toMatch(provinceData[random].name);
  });

  test('传入正确的 defaultCity, 会设置 city 默认值', () => {
    const provinceData = defaultLocation.map(item => item.children).flat();
    const random = Math.floor(Math.random() * provinceData.length);

    const wrapper = mount(
      <LocationSelector defaultProvince={provinceData[random].id} defaultCity={provinceData[random].children[0].id} />,
    );
    const cityElem = wrapper.find('.g-city-item-selected');
    expect(cityElem.text()).toMatch(provinceData[random].children[0].name);
  });

  test('传入 provinceValue, 会默认选中该 province, 并且完全受控', () => {
    const provinceData = defaultLocation.map(item => item.children).flat();
    const random = Math.floor(Math.random() * provinceData.length);
    const wrapper = mount(
      <LocationSelector provinceValue={provinceData[random].id} />,
    );
    const provinceElem = wrapper.find('.g-province-item');
    const newRandom = Math.floor(Math.random() * provinceElem.length);
    provinceElem.at(newRandom).simulate('click');
    wrapper.update();
    const targetElem = wrapper.find('.g-province-item-selected');
    expect(targetElem.text()).toMatch(provinceData[random].name);
  });

  test('传入 cityValue, 会默认选中该 city, 并且完全受控', () => {
    const provinceData = defaultLocation.map(item => item.children).flat();

    const wrapper = mount(
      <LocationSelector provinceValue={provinceData[0].id} cityValue={provinceData[0].children[0].id} />,
    );
    let cityElem = wrapper.find('.g-city-item-selected');
    expect(cityElem.text()).toMatch(provinceData[0].children[0].name);
    cityElem = wrapper.find('.g-city-item');
    cityElem.at(Math.floor(Math.random() * cityElem.length)).simulate('click');
    wrapper.update();
    cityElem = wrapper.find('.g-city-item-selected');
    expect(cityElem.text()).toMatch(provinceData[0].children[0].name);
  });

  test('传入 title, 会展示title', () => {
    const text = '地区选择器';
    const wrapper = shallow(
      <LocationSelector title={text} />,
    );
    expect(wrapper.find('.g-location-selector-title').text()).toBe(text);
  });

  test('可以正确传入类名和 style', () => {
    const style = {
      color: 'red',
    };
    const className = 'g-test';
    const wrapper = shallow(
      <LocationSelector style={style} className={className} />,
    );
    expect(wrapper.find('.g-location-selector').prop('className')).toMatch(className);
    expect(wrapper.find('.g-location-selector').prop('style')).toMatchObject(style);
  });

  test('传入 locationData, 会使用 locationData', () => {
    const testLocation = [{ id: 1, name: 'test', children: [{ id: 2, name: 'test1', children: [{ id: 3, name: 'test3', children: null }] }] }];
    const wrapper = mount(
      <LocationSelector locationData={testLocation} />,
    );
    const locationElem = wrapper.find('.g-location-item');
    expect(locationElem.length).toEqual(testLocation.length);

    for (let i = 0; i < locationElem.length; i++) {
      const locationItem = locationElem.at(i);
      const locationLabelElem = locationItem.find('.g-location-item-label').text();
      const provinceElem = locationItem.find('.g-province-item');
      expect(locationLabelElem).toMatch(testLocation[i].name);

      const provinceData = testLocation[i].children;
      expect(provinceElem.length).toBe(provinceData.length);
      for (let j = 0; j < provinceElem.length; j++) {
        expect(provinceElem.at(j).text()).toMatch(provinceData[j].name);
      }
    }
  });

  test('传入 locationSearchMethod, 会使用 locationSearchMethod 获取省份和区域,比 locationData 优先级更高', async () => {
    const testLocation = [{ id: 1, name: 'test', children: [{ id: 2, name: 'test1', children: [{ id: 3, name: 'test3', children: null }] }] }];
    const testLocation1 = [{ id: 1, name: 'test1', children: [{ id: 2, name: 'test2', children: [{ id: 3, name: 'test4', children: null }] }] }];

    const wrapper = mount(
      <LocationSelector
        locationData={testLocation}
        locationSearchMethod={() => testLocation1}
      />,
    );
    await waitForComponentToPaint(wrapper);
    const locationElem = wrapper.find('.g-location-item');

    for (let i = 0; i < locationElem.length; i++) {
      const locationItem = locationElem.at(i);
      const locationLabelElem = locationItem.find('.g-location-item-label').text();
      const provinceElem = locationItem.find('.g-province-item');
      expect(locationLabelElem).toMatch(testLocation1[i].name);

      const provinceData = testLocation1[i].children;
      expect(provinceElem.length).toBe(provinceData.length);
      for (let j = 0; j < provinceElem.length; j++) {
        expect(provinceElem.at(j).text()).toMatch(provinceData[j].name);
      }
    }
  });

  test('传入 citySearchMethod, 会使用 citySearchMethod 获取省份和区域,比 locationData 优先级更高', async () => {
    const testCity = [{ id: 1, name: 'test', children: [{ id: 2, name: 'test1', children: [{ id: 3, name: 'test3', children: null }] }] }];

    const wrapper = mount(
      <LocationSelector citySearchMethod={() => testCity} />,
    );
    const provinceElem = wrapper.find('.g-province-item');
    const random = Math.floor(Math.random() * provinceElem.length);
    const targetElem = provinceElem.at(random);
    targetElem.simulate('click');
    await waitForComponentToPaint(wrapper);
    const cityElem = wrapper.find('.g-city-item');
    expect(cityElem.length).toBe(testCity.length);

    for (let i = 0; i < cityElem.length; i++) {
      expect(cityElem.at(i).text()).toMatch(testCity[i].name);
    }
  });

  test('测试完成，保存快照', () => {
    const wrapper = shallow(
      <LocationSelector />,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
