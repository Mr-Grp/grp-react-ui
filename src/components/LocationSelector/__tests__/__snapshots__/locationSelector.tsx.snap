// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocationSelector 测试 测试完成，保存快照 1`] = `
<div
  className="g-location-selector "
  style={Object {}}
>
  <div
    className="g-location-selector-title"
  >
    地区选择器
  </div>
  <div
    className="g-location"
  />
</div>
`;
